src/main/java/com/teragrep/rlp_03/context/frame/access/Lease.java

```diff
@@ -1,28 +1,31 @@
 package com.teragrep.rlp_03.context.frame.access;
 
+import java.util.concurrent.Phaser;
+
 public class Lease implements AutoCloseable {
 
     private final Access access;
-
-    private volatile boolean isOpen;
+    private final Phaser subPhaser;
     Lease(Access access) {
         this.access = access;
-        this.isOpen = true;
+        this.subPhaser = new Phaser(1);
     }
 
 
     @Override
     public void close() {
-        if (!isOpen) {
-            throw new IllegalStateException();
+        if (subPhaser.isTerminated()) {
+            throw new IllegalStateException("lease phaser was terminated, can't close");
         }
-        else {
-            isOpen = false;
+        else if (isOpen()) {
+            subPhaser.arriveAndDeregister(); // should terminate
             access.release(this);
+        } else {
+            throw new IllegalStateException("lease not open, can't close");
         }
     }
 
     public boolean isOpen() {
-        return isOpen;
+        return !subPhaser.isTerminated();
     }
```

Comment:
I think `isTerminated()` should be used instead of inversing returns like it was used few lines above

Reply:
changed to isTerminated and changed all usages to be negated: isOpen -> !isTerminated.